html/webappapis/scripting/processing-model-2/integration-with-the-javascript-job-queue/promise-job-entry.html

```diff
@@ -0,0 +1,99 @@
+<!-- https://github.com/whatwg/html/issues/1426 -->
+
+<!-- This is the entry page, so window.open() should resolve relative to it, even inside promise jobs. -->
+
+<iframe src="resources/promise-job-entry-incumbent.html"></iframe>
+
+<script>
+setup({ explicit_done: true });
+
+const relativeURL = "resources/window-to-open.html";
+const expectedURL = (new URL(relativeURL, location.href)).href;
+
+window.onload = () => {
+  async_test(t => {
+    const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+    w.onload = t.step_func_done(() => {
+      t.add_cleanup(() => w.close());
+      assert_equals(w.location.href, expectedURL);
+    });
+  }, "Sanity check: this all works as expected with no promises involved");
+
+  async_test(t => {
+    // No t.step_func because that could change the realms
+    Promise.resolve().then(() => {
+      const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+      w.onload = t.step_func_done(() => {
+        t.add_cleanup(() => w.close());
+        assert_equals(w.location.href, expectedURL);
+      });
+    });
+  }, "Fulfillment handler on fulfilled promise");
+
+  async_test(t => {
+    // No t.step_func because that could change the realms
+    Promise.reject().catch(() => {
+      const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+      w.onload = t.step_func_done(() => {
+        t.add_cleanup(() => w.close());
+        assert_equals(w.location.href, expectedURL);
+      });
+    });
+  }, "Rejection handler on rejected promise");
+
+  async_test(t => {
+    let resolve;
+    const p = new Promise(r => { resolve = r; });
+
+    // No t.step_func because that could change the realms
+    p.then(() => {
+      const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+      w.onload = t.step_func_done(() => {
+        t.add_cleanup(() => w.close());
+        assert_equals(w.location.href, expectedURL);
+      });
+    });
+
+    t.step_timeout(resolve, 0);
+  }, "Fulfillment handler on pending-then-fulfilled promise");
+
+  async_test(t => {
+    let reject;
+    const p = new Promise((_, r) => { reject = r; });
+
+    // No t.step_func because that could change the realms
+    p.catch(() => {
+      const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+      w.onload = t.step_func_done(() => {
+        t.add_cleanup(() => w.close());
+        assert_equals(w.location.href, expectedURL);
+      });
+    });
+
+    t.step_timeout(reject, 0);
+  }, "Rejection handler on pending-then-rejected promise");
+
+  async_test(t => {
+    const thenable = {
+      // No t.step_func because that could change the realms
+      then(f) {
+        const w = frames[0].runWindowOpenVeryIndirectly(relativeURL);
+        w.onload = t.step_func_done(() => {
+          t.add_cleanup(() => w.close());
+          assert_equals(w.location.href, expectedURL);
+        });
+      }
+    };
+
+    Promise.resolve(thenable);
+  }, "Thenable resolution");
+
+  done();
+};
+</script>
```

Comment:
Thank you, this is a great start.  Additional things worth testing:

1) When the Promise comes from one global but the function passed to `then` or `catch` comes from a different global.
2) When the Promise comes from one global, the thenable passed to `resolve` from another global (well, maybe it's OK if it's the same one), and the `then` function on the thenable comes from a different global.
3) Whether browsers do the "can run a script" checks based one various entry globals.

Followup is fine for adding those; the third one especially is likely to be a bit of a pain.

Reply:
Added (1) and (2)

Reply:
Filed https://github.com/web-platform-tests/wpt/issues/21756 for (3); thoughts appreciated.